mount: handle loopback mount

bergwolf opened the pull request:
If a mount has specified `loop` option, we need to handle it on our
own instead of passing it to the kernel. In such case, create a
loopback device, attach the mount source to it, and mount the loopback
device rather than the mount source.

This is a spit from #3130 so that rawblock can be a non-builtin snapshotter.

/cc @dmcgowan @AkihiroSuda

AkihiroSuda commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,178 @@
+func setupLoop(backingFile string, ro bool) (*os.File, error) {
+	var loopDev string
+
+	for retry := 1; ; retry++ {
```
max retry count?

stevvooe commented:
Would it be possible to reconcile this code with the other losetup code in containerd? https://github.com/containerd/containerd/search?q=losetup&unscoped_q=losetup provides some references. I didn't look into how similar they are.

bergwolf commented:
@stevvooe I've implemented the same losetup APIs needed by `devmapper`, and replaced its private losetup command line wrapper with `mount/losetup`. PTAL.

bergwolf commented:
btw, the failed ci looks unrelated:
```
##[error]github.com/containerd/containerd/api/events/container.proto:6:1: warning: Import gogoproto/gogo.proto is unused.
##[error]github.com/containerd/containerd/api/events/namespace.proto:5:1: warning: Import gogoproto/gogo.proto is unused.
panic: protobuf tag not enough fields in FileDescriptorSet.state: 
```

estesp commented:
We had a problem with unpinned dependencies breaking our protobuf checks in CI; if you can rebase on master HEAD, you should get a clean CI run now.  Thanks!

mxpv commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,229 @@
+			// with EBUSY when trying to set it up.
+			if strings.Contains(err.Error(), ebusyString) {
+				// Fallback a bit to avoid live lock
+				time.Sleep(time.Millisecond * time.Duration(rand.Intn(retry*10)))
```
If I get this logic correctly, this func may sleep way too long:
```go
	total := time.Duration(0)
	for retry := 1; retry < 10000; retry++ {
                // time.Sleep(...)
		total += time.Millisecond * time.Duration(rand.Intn(retry*10))
	}
	fmt.Printf("%f", total.Hours()) // Prints 70 hours.
```

bergwolf replied:
I was trying to follow the logic in `losetup.c` (https://github.com/karelzak/util-linux/blob/master/sys-utils/losetup.c#L548) where it retries forever on `EBUSY`. @AkihiroSuda suggested using a max retry count so I added this extremely large retry.

mxpv replied:
I'd propose some reasonable sleep times here. Like if we can't do it within 1 min, then just fail. @AkihiroSuda WDYT?

bergwolf replied:
nvm. I've changed it to retry at most 200 times that can translate into a little more than 1 minute.

mxpv commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return err
+	}
+	_, _, err = ioctl(dev.Fd(), ioctlClrFd, 0)
```
Do we need to close here?

bergwolf replied:
Good catch, thanks!

mxpv commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,229 @@
+	if param.Direct {
+		info.flags |= loFlagsAutoclear
+	}
+	if _, _, err := ioctl(loopFile.Fd(), ioctlSetStatus64, uintptr(unsafe.Pointer(&info))); err != nil {
```
What's the benefit of having this losetup code vs the old package ?

bergwolf replied:
The main purpose of the PR is to enable `mount` package to handle loopback mounts so that I can use it for the rawblock snapshotter. Maintainers also want to consolidate the two losetup implementation so I added the same APIs to replace the `losetup` command line wrappers. The main difference between them is calling the `losetup` binary or not.

mxpv replied:
Sorry, I meant why not just reuse existing package. Potentially it may be a lot harder to maintain and support, since Go is not the best fit for low level programming.

bergwolf replied:
Because it is not possible to create a autoclear loopback device with `losetup` command line. The workflow I want is:
1. open the loop device
2. set its backend and set the `autoclear` flag
3. mount the loopback device with proper file system
4. close the initial open file descriptor that we got in step 1

This is basically what you get with `mount -o loop` but it is not possible with the `losetup` command line. And the benefit is worthwhile -- we don't have to care about deleting the loopback device after unmounting the file system, nor if anything goes wrong in the middle before we mount the file system. The kernel takes care of cleaning up the loopback device once the last opener to it closes its open file descriptor.

bergwolf replied:
And we're already calling mount syscall instead of the `mount` binary in the mount package. Doing the same for loopback device follows the same philosophy.

mxpv replied:
> Because it is not possible to create an autoclear loopback device with losetup command line.

hm. As far as I understand `losetup` does that with `--detach` flag.

```
-d, --detach loopdev...
              Detach the file or device associated with the specified loop
              device(s). Note that since Linux v3.7 kernel uses "lazy device
              destruction".  The detach operation does not return EBUSY
              error anymore if device is actively used by system, but it is
              marked by autoclear flag and destroyed later.
```

http://man7.org/linux/man-pages/man8/losetup.8.html

bergwolf replied:
It is the detach action, not setting a loop device as autoclear. With autoclear, the loop device is detached automatically **after** it is unmounted -- that means we do not need to call `losetup -d` after unmounting a loop device.

If we use `losetup` command line, the workflow is like:
1. `losetup -f <backing-file>`
2. mount <loop-device> with a file system
3. unmount the file system
4. `losetup -d <loop-device>`

Whereas with autoclear, we do:
1. `mount -o loop <backing-file>` alike with `mount` package
2. umount the file system

bergwolf replied:
Or do you mean that we should do the following?
1. `losetup -f <backing-file>`
2. mount with a file system
3. `losetup -d <loop-device>`
4. unmount the file system

That still doesn't look as neat as just doing
1. `mount -o loop <backing-file>` alike with mount package
2. umount the file system

bergwolf commented:
@estesp Thanks for the hint. I've rebased on top of master.

zhsj commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+// struct loop_info64 in util-linux/include/loopdev.h
+type loopInfo struct {
```
I think you can use https://godoc.org/golang.org/x/sys/unix#LoopInfo

zhsj commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,230 @@
+	_ [112]byte
+}
+
+func ioctl(fd, req, args uintptr) (uintptr, uintptr, error) {
```
maybe using Ioctl* family from golang.org/x/sys/unix?

zhsj commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,230 @@
+	ioctlSetFd       = 0x4C00
+	ioctlClrFd       = 0x4C01
+	ioctlSetStatus64 = 0x4C04
+	ioctlGetFree     = 0x4C82
```
unix.LOOP_SET_FD
unix.LOOP_CLR_FD
unix.LOOP_SET_STATUS64
unix.LOOP_CTL_GET_FREE

zhsj commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,230 @@
+	//loFlagsUseAops   = 2
+	loFlagsAutoclear = 4
+	//loFlagsPartScan  = 8
+	loFlagsDirectIO = 16
```
unix.LO_FLAGS_READ_ONLY
unix.LO_FLAGS_AUTOCLEAR
unix.LO_FLAGS_PARTSCAN
unix.LO_FLAGS_DIRECT_IO

zhsj commented on mount/losetup_linux.go:
```diff
@@ -0,0 +1,230 @@
+	//loFlagsPartScan  = 8
+	loFlagsDirectIO = 16
+
+	ebusyString = "device or resource busy"
```
unix.EBUSY.Error() is this. But I think you don't need to compare the error string.

mxpv commented:
+1 for using existing routines from `unix` package. Could you possibly address that @bergwolf ?

AkihiroSuda commented:
ping @bergwolf

mxpv commented:
Carried in #4902
